model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java

```diff
@@ -247,7 +248,7 @@ protected void checkRemoteCaches(KeycloakSession session) {
         Cache<String, SessionEntityWrapper<UserSessionEntity>> sessionsCache = ispn.getCache(InfinispanConnectionProvider.USER_SESSION_CACHE_NAME);
         boolean sessionsRemoteCache = checkRemoteCache(session, sessionsCache, (RealmModel realm) -> {
             // We won't write to the remoteCache during token refresh, so the timeout needs to be longer.
-            return realm.getSsoSessionMaxLifespan() * 1000;
+            return Time.currentTimeMillis() + (realm.getSsoSessionMaxLifespan() * 1000L);
```

Comment:
I don't think that adding "Time.currentTimeMillis()" is correct here?

This is idle timeout. Hence if "Time.currentTimeMillis" is added, it means the expiration won't be around 10 hours (as default SSO session max lifespan), but around 40+ years (Time since 1 Jan 1970). Use something like `(1000L * realm.getSsoSessionMaxLifespan()` should do the trick with retype to long though?

Reply:
Hi Marek! No problems :)

Actually, when a remote Infinispan cache is set up the model expects an epoch, not a lifespan, this is why adding Time.currentTimeMillis solves the issue. I'm using `org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory` to marshall data from keycloak to hot rod server

Reply:
Hi Alexandre,

I am not sure how you figured this statement? When looking at the javadoc of method org.infinispan.commons.api.BasicCache and method `put(K key, V value, long lifespan, TimeUnit lifespanUnit, long maxIdleTime, TimeUnit maxIdleTimeUnit)` , it mentions this for the maxIdleTime field: the maximum amount of time this key is allowed to be idle for before it is considered as expired .

This is the method, which Keycloak internally calls. So I don't understand why this is supposed to be the epoch? Any more info about this?

Reply:
I've setup my environment the like described below:

1 - An external Infinispan Hot Rod Server with JDBC store configuration
```
Configuration persistentCacheConfiguration = new ConfigurationBuilder()
                    .memory().size(100).evictionStrategy(EvictionStrategy.REMOVE)
                    .persistence().addStore(JdbcStringBasedStoreConfigurationBuilder.class)
                    .table()
                    .idColumnName("id")
                    .idColumnType("varchar(1000)")
                    .dataColumnName("data")
                    .dataColumnType("bytea")
                    .timestampColumnName("expire")
                    .timestampColumnType("bigint")
                    .dropOnExit(false)
                    .createOnStart(true)
                    .tableNamePrefix("ispn")
                    .connectionPool()
                    .connectionUrl(cfg.get("DATABASE_JDBC_URL"))
                    .driverClass(cfg.get("DATABASE_JDBC_DRIVER"))
                    .password(cfg.get("DATABASE_PASSWORD"))
                    .username(cfg.get("DATABASE_USERNAME"))
                    .build();

            DefaultCacheManager defaultCacheManager = new DefaultCacheManager(
                    new GlobalConfigurationBuilder()
                            .defaultCacheName("sessions")
                            .build(), persistentCacheConfiguration);

            defaultCacheManager.createCache("___script_cache", new ConfigurationBuilder().build());
            defaultCacheManager.createCache("clientSessions", persistentCacheConfiguration);
            defaultCacheManager.createCache("offlineClientSessions", persistentCacheConfiguration);
            defaultCacheManager.createCache("offlineSessions", persistentCacheConfiguration);
            defaultCacheManager.createCache("actionTokens", persistentCacheConfiguration);
            defaultCacheManager.createCache("loginFailures", persistentCacheConfiguration);

            new HotRodServer().start(new HotRodServerConfigurationBuilder().host("127.0.0.1").port(11322).build(),
                    defaultCacheManager);
```

2 - Remote socket configuration
```
/socket-binding-group=standard-sockets/remote-destination-outbound-socket-binding=remote-cache:add(host="127.0.0.1", port="11322")
```

3 - Cache configuration

```
/subsystem=infinispan/cache-container=keycloak/local-cache=sessions/store=remote:add(cache="sessions", remote-servers=[remote-cache], preload=false, purge=false, passivation=false, shared=false, properties={rawValues=true, marshaller=org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory, protocolVersion=${env.HOTROD_VERSION:2.8}}){allow-resource-service-restart=true}
/subsystem=infinispan/cache-container=keycloak/local-cache=clientSessions/store=remote:add(cache="clientSessions", remote-servers=[remote-cache], preload=false, purge=false, passivation=false, shared=false, properties={rawValues=true, marshaller=org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory, protocolVersion=${env.HOTROD_VERSION:2.8}}){allow-resource-service-restart=true}
/subsystem=infinispan/cache-container=keycloak/local-cache=offlineClientSessions/store=remote:add(cache="offlineClientSessions", remote-servers=[remote-cache], preload=false, purge=false, passivation=false, shared=false, properties={rawValues=true, marshaller=org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory, protocolVersion=${env.HOTROD_VERSION:2.8}}){allow-resource-service-restart=true}
/subsystem=infinispan/cache-container=keycloak/local-cache=offlineSessions/store=remote:add(cache="offlineSessions", remote-servers=[remote-cache], preload=false, purge=false, passivation=false, shared=false, properties={rawValues=true, marshaller=org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory, protocolVersion=${env.HOTROD_VERSION:2.8}}){allow-resource-service-restart=true}
/subsystem=infinispan/cache-container=keycloak/local-cache=authenticationSessions/store=remote:add(cache="authenticationSessions", remote-servers=[remote-cache], preload=false, purge=false, passivation=false, shared=false, properties={rawValues=true, marshaller=org.keycloak.cluster.infinispan.KeycloakHotRodMarshallerFactory, protocolVersion=${env.HOTROD_VERSION:2.8}}){allow-resource-service-restart=true}
```

After performing a login on Keycloak, I retrieved some data from database table ispn_clientSessions. There the timestamp field is not a max idle but an epoch